### DISCUSSION TITLE
Highlighting substructure matches based on elemental composition only

### DISCUSSION THREAD

**Discussion by jpfeuffer** [rdkit/rdkit#6547] [Q&A] [answered]
Hi!

I was wondering if there is any functionality to search for all substructures in a molecule that match a given elemental composition (no other connectivity constraints).
My substructures are fragments of the molecule (an alternative would certainly be to generate all possible fragments).
But I only have the masses of those fragments and through computation a possible elemental composition.
I am aware that it can be computationally complex but maybe someone has a good idea on how to do this with RDKit.
The goal, for now, is just to highlight the fragments in the parent molecule.

Thanks!

**Comment by ptosco** [comment, marked as answer]
This is something rather crude and inefficient (and barely tested) that I have drafted while commuting home, but it should be enough to get you started.
Note that if your goal is to relate MS peaks to molecule substructures, a large number of those substructures, though formally correct, would not actually be obtainable through any viable fragmentation scheme.
```
from rdkit import Chem
from rdkit.Chem.Draw import MolsToGridImage

mol = Chem.MolFromSmiles("CC(C)N(C)C(C)c1cnc(C)nc1")
mol
```
![image](https://github.com/rdkit/rdkit/assets/5244385/7c54347d-4636-419f-b239-70a3d5ec6945)

```
class SubgraphByElementalComposition:
    def __init__(self, mol, composition):
        self.mol = mol
        self.composition = composition
        self.n_elements = sum(self.composition.values())

    def filter_subgraph_by_composition(self, subgraph):
        counts = {}
        ai_subgraph, _ = subgraph
        for ai in ai_subgraph:
            atom = self.mol.GetAtomWithIdx(ai)
            k = atom.GetSymbol()
            max_count = self.composition.get(k, None)
            if max_count is None:
                return False
            count = counts.get(k, 0) + 1
            if count > max_count:
                return False
            counts[k] = count
        return counts == self.composition

    def filter_subgraphs_by_composition(self, subgraphs):
        return filter(self.filter_subgraph_by_composition, subgraphs)

    def bond_indices_to_atom_bond_indices(self, bond_indices):
        atom_indices = {self.mol.GetBondWithIdx(bi).GetBeginAtomIdx() for bi in bond_indices}
        atom_indices.update(self.mol.GetBondWithIdx(bi).GetEndAtomIdx() for bi in bond_indices)
        return atom_indices, bond_indices
 
    def get_subgraphs_with_composition(self):
        min_edges = self.n_elements - 1
        max_edges = self.n_elements * (self.n_elements - 1) // 2
        subgraphs = sum(list(Chem.FindAllSubgraphsOfLengthMToN(self.mol, min_edges, max_edges)), [])
        subgraphs = [self.bond_indices_to_atom_bond_indices(bond_indices) for bond_indices in subgraphs]
        return self.filter_subgraphs_by_composition(subgraphs)
        
    @classmethod
    def get(cls, mol, composition):
        instance = cls(mol, composition)
        return instance.get_subgraphs_with_composition()

subgraphs = list(SubgraphByElementalComposition.get(mol, {"C": 5, "N": 3}))
atom_lists = [list(atom_indices) for atom_indices, _ in subgraphs]
bond_lists = [list(bond_indices) for _, bond_indices in subgraphs]
MolsToGridImage([Chem.Mol(mol) for _ in range(len(bond_lists))], highlightAtomLists=atom_lists, highlightBondLists=bond_lists, molsPerRow = 5) if bond_lists else None
```
![image](https://github.com/rdkit/rdkit/assets/5244385/6712e5e5-9667-4be3-98e3-3ab0d48f86ee)

**Reply by jpfeuffer** [reply]
Hi @ptosco 

Thank you very much for your very quick reply!

This is great and really got me started.
Indeed it's for matching MS fragment peaks. However, since the fragmentation seems not to be fully understood and varies a lot between techniques and collision energies I thought it would be a good idea to avoid missing potential matches.
From your example, though, I can see already that I underestimated the amount of matches.

I will probably follow an in-silico fragmentation approach now.

I have a quick general question regarding that:
If I use fragmentAtBonds for example, will the generated Molecule objects be copies of the parent Molecule or views on it, just storing a subset of the atoms?

**Comment by ptosco** [comment]
If you are referring to [FragmentOnBonds](https://www.rdkit.org/docs/source/rdkit.Chem.rdmolops.html#rdkit.Chem.rdmolops.FragmentOnBonds), it returns a new molecule, leaving the input molecule unaltered.

**Reply by jpfeuffer** [reply]
Ok! Would there be a way to force a view/pointer on the original molecule?
During recursive fragmentation along the fragmentation tree I assume this otherwise would be a bit wasteful in memory.
Otherwise I would probably just follow a bitarray approach to keep track of the atoms and bonds in each resulting fragment.
